### PR TITLE
Feature/ike freetype optimize

### PR DESCRIPTION
**Pull request by IkeBart**
This branch should remove some calls to FT_Load_Glyph() when using MOAIFreeTypeFont::_renderTexture() and reduce the amount of time needed to execute that method.

TODO:
- [x] test in Instruments.

### CODE REVIEW

File: src/moaicore/MOAIFreeTypeFont.cpp [hunk 1]
```diff
@@ -1059,13 +1067,14 @@ int MOAIFreeTypeFont::NumberOfLinesToDisplayText(cc8 *text, FT_Int imageWidth,
 	int n = 0;
 	while ( (unicode = u8_nextchar(text, &n) ) ) {
 		
-		startIndex = (u32) lineIndex;
+		startIndex = (u32) lineIndex; //( (int)glyphArrayIndex - rewindCount); //lineIndex;
```
Comment (kstenerud) [inline src/moaicore/MOAIFreeTypeFont.cpp]: Commented out stuff

File: src/moaicore/MOAIFreeTypeFont.cpp [hunk 2]
```diff
@@ -1059,13 +1067,14 @@ int MOAIFreeTypeFont::NumberOfLinesToDisplayText(cc8 *text, FT_Int imageWidth,
...
 		
 		// handle new line
 		if (unicode == '\n'){
 			numberOfLines++;
 			penX = penXReset;
-			lineIndex = tokenIndex = n;
+			lineIndex = tokenIndex = glyphArrayIndex; //n - 1;
```
Comment (kstenerud) [inline src/moaicore/MOAIFreeTypeFont.cpp]: Commented out stuff

File: src/moaicore/MOAIFreeTypeFont.cpp [hunk 3]
```diff
@@ -1078,7 +1087,8 @@ int MOAIFreeTypeFont::NumberOfLinesToDisplayText(cc8 *text, FT_Int imageWidth,
 		}
 		// handle word breaking characters
 		else if ( MOAIFreeTypeFont::IsWordBreak(unicode, wordBreakMode) ){
-			tokenIndex = n;
+			tokenIndex = glyphArrayIndex; //n;
```
Comment (kstenerud) [inline src/moaicore/MOAIFreeTypeFont.cpp]: Commented out stuff

File: src/moaicore/MOAIFreeTypeFont.cpp [hunk 4]
```diff
@@ -1117,7 +1143,7 @@ int MOAIFreeTypeFont::NumberOfLinesToDisplayText(cc8 *text, FT_Int imageWidth,
 				numberOfLines++;
 				textLength = 0;
 				penX = penXReset;
-				lineIndex = tokenIndex = n;
+				lineIndex = tokenIndex = glyphArrayIndex; //n - 1;
```
Comment (kstenerud) [inline src/moaicore/MOAIFreeTypeFont.cpp]: Commented out stuff

File: src/moaicore/MOAIFreeTypeFont.cpp [hunk 5]
```diff
@@ -1141,9 +1167,10 @@ int MOAIFreeTypeFont::NumberOfLinesToDisplayText(cc8 *text, FT_Int imageWidth,
 						}
 						
 					}
-					
+					// set the rewind count for skipping glyphs already loaded.
+					rewindCount =  (glyphArrayIndex - tokenIndex) - 2; //(n - tokenIndex) - 1;
```
Comment (kstenerud) [inline src/moaicore/MOAIFreeTypeFont.cpp]: Commented out stuff

File: src/moaicore/MOAIFreeTypeFont.cpp [hunk 6]
```diff
@@ -1155,7 +1182,7 @@ int MOAIFreeTypeFont::NumberOfLinesToDisplayText(cc8 *text, FT_Int imageWidth,
 					//advance to next line
 					numberOfLines++;
 					penX = penXReset;
-					lineIndex = tokenIndex = n;
+					lineIndex = tokenIndex = glyphArrayIndex - (rewindCount + 1); //n - 1;
```
Comment (kstenerud) [inline src/moaicore/MOAIFreeTypeFont.cpp]: Commented out stuff

File: src/moaicore/MOAIFreeTypeFont.cpp [hunk 7]
```diff
@@ -1169,7 +1196,7 @@ int MOAIFreeTypeFont::NumberOfLinesToDisplayText(cc8 *text, FT_Int imageWidth,
 						// advance to next line
 						numberOfLines++;
 						penX = penXReset;
-						lineIndex = tokenIndex = n;
+						lineIndex = tokenIndex = glyphArrayIndex; //n - 1;
```
Comment (kstenerud) [inline src/moaicore/MOAIFreeTypeFont.cpp]: Commented out stuff

File: src/moaicore/MOAIFreeTypeFont.cpp [hunk 8]
```diff
@@ -1500,6 +1534,12 @@ MOAITexture* MOAIFreeTypeFont::RenderTexture(cc8 *text, float size, float width,
 	MOAITexture *texture = new MOAITexture();
 	texture->Init(bitmapImg, "");
 	
+	// clean up the glyph array
+	deleteGlyphArray(this->mGlyphArray, glyphArraySize);
+	
+	// clean up the advance array
+	delete [] this->mAdvanceArray;
```
Comment (kstenerud) [inline src/moaicore/MOAIFreeTypeFont.cpp:1539]: Are these pointers safe to leave without setting to NULL?

File: src/moaicore/MOAIFreeTypeFont.h [hunk 9]
```diff
@@ -62,6 +62,13 @@ class MOAIFreeTypeFont : public virtual MOAILuaObject {
 	u32 mBitmapHeight;
 	
 	std::vector<MOAIFreeTypeTextLine> mLineVector;
+	
+	FT_Glyph* mGlyphArray;
+	FT_Vector* mAdvanceArray;
+	
+	// A member variable used for determining which call to a method generated an error.
+	// TODO: remove it in final product
```
Comment (kstenerud) [inline src/moaicore/MOAIFreeTypeFont.h]: TODO reminder

**Comment by kstenerud** [comment]
:shipit:

File: src/moaicore/MOAIFreeTypeFont.cpp [hunk 10]
```diff
@@ -1046,11 +1048,16 @@ int MOAIFreeTypeFont::NumberOfLinesToDisplayText(cc8 *text, FT_Int imageWidth,
 	int lineIndex = 0; // the index of the beginning of the current line
 	int tokenIndex = 0; // the index of the beginning of the current token
 	
+	int tokenN = 0; // the last value of n at the beginning of the token.
```
Comment (NachoSoto) [inline src/moaicore/MOAIFreeTypeFont.cpp:1051]: Maybe this variable can have a better name so that you don't need a comment? :)

**Comment by NachoSoto** [comment]
:shipit: once you've verified there are no leaks.

**Comment by aab29** [comment]
Isaac says: "In my profile tests, it looks like there is an improvement when using the exact same setup across branches.  I compared the 'feature/ike-freetype-optimize' and 'mindsnacks-dev' branches.  It took between fifteen and twenty percent less time to run MOAIFreeTypeFont::_render() on 'feature/ike-freetype-optimize' branch than on the 'mindsnacks-dev' branch in the two setups I used.  The larger improvement went to setup with the smaller font size."

:shipit:

**Comment by NachoSoto** [comment]
@aab29 have we verified we haven't introduced any leaks though? /cc @IkeBart

**Comment by NachoSoto** [comment]
Bump @aab29

**Comment by aab29** [comment]
Isaac profiled the thing quite a bit.  I'm guessing he would have noted any leaks.  Were there any leaks?  @IkeBart

**Comment by NachoSoto** [comment]
The Time Profiler is different from the Leaks instrument
screen shot 2014-02-24 at 10 54 28 am
He didn't say anything about it.

**Comment by NachoSoto** [comment]
Bump. I gave this my :shipit:  **once** we had verified no new leaks were introduced.

**Comment by kstenerud** [comment]
Says it's already been merged?

**Comment by NachoSoto** [comment]
That's exactly my point, we merged this before checking that.

**Comment by IkeBart** [comment]
I did a test with the Leaks instrument with a Lua program that runs MOAITextRenderer::_render() five thousand times, and found nothing on the leaks instrument.

**Comment by kstenerud** [comment]
review

**Comment by NachoSoto** [comment]
@kstenerud I wrote that comment before we merged this, that's why I'm insisting to know if we had checked or not.

**Comment by NachoSoto** [comment]
Following up: https://github.com/mindsnacks/moai-dev/pull/64

**Pull request merged**